### PR TITLE
EMAC API reference

### PR DESCRIPTION
**Pull request by mikaleppanen**
Added EMAC and EMAC Memory Manager API descriptions under new "\reference\api\connectivity\drivers" directory. 

Please review @SeppoTakalo @AnotherButler 

## NOTE: DO NOT MERGE TO 5.8. THIS IS FOR 5.9 ONLY

### CODE REVIEW

**Comment by SeppoTakalo** [comment]
@AnotherButler 
Please note that this is also targeting 5.9. Should not be merged to 5.8

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+If your EMAC driver is the default for the target, define the static get default instance function to return an instance of your EMAC.
+
+### Usage
+
+The EMAC driver class is instantiated during the creation of the network interface. When the network interface is brought up, the network stack powers the EMAC driver.
+
+Steps that the network stack uses to power the EMAC driver:
+
+1. The EMAC memory manager class reference is configured to the driver.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what configures the EMAC memory manager class reference to the driver?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+If your EMAC driver is the default for the target, define the static get default instance function to return an instance of your EMAC.
+
+### Usage
+
+The EMAC driver class is instantiated during the creation of the network interface. When the network interface is brought up, the network stack powers the EMAC driver.
+
+Steps that the network stack uses to power the EMAC driver:
+
+1. The EMAC memory manager class reference is configured to the driver. 
+1. The link input and state callbacks are set.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what sets the link input and state callbacks?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+1. The driver powers up.
+1. MTU and hardware address sizes are read from the driver.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what reads the MTU and hardware address sizes from the driver?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+1. Hardware address is read from the driver.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what writes the hardware address to the driver?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+1. Hardware address is written to driver.
+    1. If the driver does not provide address on the read call, the network stack chooses the address from the central system configuration and writes that to driver.
+
+Optional steps:
+
+1. Interface name is read from the driver.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what reads the interface name from the driver?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+1. Multicast filtering is configured.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what configures multicast filtering?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    1. The driver can either support multicast filtering or provide all frames.
+1. The memory buffer align preference is read from the driver.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Query: Who or what reads the memory buffer align preference from the driver?

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
+## EMAC
+
+The Ethernet MAC device driver class provides an interface for a network stack to control the Etherner driver.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: When would you want to use this API and why? What are its use cases?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
+## EMAC memory manager
+
+The Ethernet MAC memory manager class provides abstracted memory interface toward memory modules used in different network stacks. The EMAC device driver uses the memory interface to handle memory related operations, such as allocation, free and data manipulation.
+
+Memory buffer chains store the data on the memory interface. The data passed in either direction either may be contiguous (a single-buffer chain) or may consist of multiple buffers. You can chain the buffers using a singly linked list.
+
+### Usage
+
+On link output to EMAC driver, buffer chain ownership is given to the driver. Driver must free the buffer after it is no longer used. The driver can express alignment preferences for buffers sent to it, but the network stack is not required to meet these preferences. This means that the drivers relying on alignment, may need to copy the output data into an aligned (or contiguous) buffers.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Who or what gives buffer chain ownership to the driver?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+EMAC driver can allocate memory either from a memory pool or from the heap (contiguous memory). By preference, link input memory should be allocated from the pool, but if contiguous memory is required it can be allocated from the heap.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Who or what should allocate link input memory from the pool? You, the user? Also, who or what might require contiguous memory? If contiguous memory's necessary, who or what can allocate link input memory from the heap?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+Memory buffer chains store the data on the memory interface. The data passed in either direction either may be contiguous (a single-buffer chain) or may consist of multiple buffers. You can chain the buffers using a singly linked list.
+
+### Usage
+
+On link output to EMAC driver, buffer chain ownership is given to the driver. Driver must free the buffer after it is no longer used. The driver can express alignment preferences for buffers sent to it, but the network stack is not required to meet these preferences. This means that the drivers relying on alignment, may need to copy the output data into an aligned (or contiguous) buffers.
+
+EMAC driver can allocate memory either from a memory pool or from the heap (contiguous memory). By preference, link input memory should be allocated from the pool, but if contiguous memory is required it can be allocated from the heap. 
+
+On link input to the network stack, buffer chain ownership is given to the stack. Stack frees the buffer after it no longer uses it.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Who or what gives buffer chain ownership to the stack?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 12]
```diff
@@ -0,0 +1,31 @@
+## EMAC memory manager
+
+The Ethernet MAC memory manager class provides abstracted memory interface toward memory modules used in different network stacks. The EMAC device driver uses the memory interface to handle memory related operations, such as allocation, free and data manipulation.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: This is good. I also want to know who would use this and when.

**Comment by SeppoTakalo** [comment]
* Clarified buffer responsibilities
* Few other clarifications on who calls what
* Remove examples, these belong to the Porting Guide
* Remove TBDs

**Comment by SeppoTakalo** [comment]
@AnotherButler Please review.

File: docs/reference/api/connectivity/drivers/EMAC.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+1. The network stack might query the interface name from the driver.
+1. The network stack might configure multicast filtering.
+    1. The driver can either support multicast filtering or provide all frames.
+1. The network stack might query for the memory buffer align preference from the driver.
+    1. The network stack is not required to use the alignment for the memory buffers on link out.
+
+### EMAC class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/class_emac.html)
+
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMAC.md]: Please include a transcluded example from the teams repo to match our user API template.

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+The driver can express alignment preferences for buffers sent to it, but the network stack is not required to meet these preferences. This means that the drivers relying on alignment need to check the buffer address and may need to copy the output data into an aligned (or contiguous) buffer when its requirements are not met.
+
+On link output to the EMAC driver, buffer chain ownership is given to the driver from network stack. The driver must free the buffer after it is no longer used.
+
+On link input to the network stack, buffer chain ownership is given to the stack from driver. Stack frees the buffer after it no longer uses it.
+
+### EMAC memory manager class reference
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/class_emac_memory_manager.html)
+
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Please include a transcluded example from the teams repo to match our user API template.

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 15]
```diff
@@ -0,0 +1,24 @@
+## EMAC memory manager
+
+The Ethernet MAC memory manager class provides abstracted memory interface toward memory modules used in different network stacks. The EMAC device driver uses the memory interface to handle memory related operations, such as allocation, free and data manipulation. The [EMAC](emac.html) drivers use this memory manager to reserve space for network packets.
+
+Memory buffer chains store the data on the memory interface. The data passed in either direction either may be contiguous (a single-buffer chain) or may consist of multiple buffers. You can chain the buffers using a singly linked list.
+
+### Usage
+
+The EMAC driver can allocate memory either from a memory pool or from the heap (contiguous memory). By preference, link input memory should be allocated from the pool, but if contiguous memory is required by the hardware, it can be allocated from the heap.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Is this saying that you, the user, should allocate link input memory from the pool unless the hardware requires contiguous memory, in which case you should allocate the link input memory from the heap?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+The driver can express alignment preferences for buffers sent to it, but the network stack is not required to meet these preferences. This means that the drivers relying on alignment need to check the buffer address and may need to copy the output data into an aligned (or contiguous) buffer when its requirements are not met.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Is this saying that drivers the rely on alignment must check the buffer address, and those drivers may need to copy the output data into an aligned (or contiguous) buffer when the buffer address's needs aren't met?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+On link output to the EMAC driver, buffer chain ownership is given to the driver from network stack. The driver must free the buffer after it is no longer used.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Who or what gives buffer chain ownership to the driver?

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+On link input to the network stack, buffer chain ownership is given to the stack from driver. Stack frees the buffer after it no longer uses it.
```
Comment (AnotherButler) [inline docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Query: Who or what gives buffer chain ownership to the stack?

File: docs/reference/api/connectivity/drivers/EMAC.md (hunk 13 above)
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/drivers/EMAC.md]: But this is not user API.

This is driver API and I feel like examples should be given in porting document, if somewhere.

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md (hunk 14 above)
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Same as previous.
This is not user API and therefore I feel we cannot be giving application examples.

Example drivers are in Mbed OS tree, and if we wan't to give code snippets, those should be in porting document.

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md (hunk 15 above)
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: If the `user` is the driver, then yes.

This is an API that is used by the driver, not Mbed OS user.

File: docs/reference/api/connectivity/drivers/EMACMemoryManager.md (hunk 16 above)
Comment (SeppoTakalo) [reply docs/reference/api/connectivity/drivers/EMACMemoryManager.md]: Yes, correct.
Feel free to change the messaging if you feel that it is not expressing its intent clearly enough.

**Comment by AnotherButler** [comment]
So we categorize our API references as:

- contributing (porting) APIs, which have porting people (usually Partners) as the target audience.
- user APIs, which have developers as the target audience.

We put the porting APIs in the contributing section and the user APIs in the API section. They have slightly different templates because they have different uses. The porting API references do not have examples but do have a testing section. Is it OK if I move these into contributing?

**Comment by SeppoTakalo** [comment]
Ah, ok.
That separation makes sense.

I'll then close this PR and move relevant sections into #498

**Comment by mikaleppanen** [comment]
Closing this.

**Pull request closed without merging**